
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>graphkb/genes - GraphKB Python</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.4b4a2bd9.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#graphkbgenes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="GraphKB Python" class="md-header__button md-logo" aria-label="GraphKB Python" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GraphKB Python
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              graphkb/genes
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/bcgsc/pori_graphkb_python" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    github
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="GraphKB Python" class="md-nav__button md-logo" aria-label="GraphKB Python" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    GraphKB Python
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/bcgsc/pori_graphkb_python" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    github
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Notice
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#get_oncokb_oncogenes" class="md-nav__link">
    get_oncokb_oncogenes()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_oncokb_tumour_supressors" class="md-nav__link">
    get_oncokb_tumour_supressors()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_cancer_genes" class="md-nav__link">
    get_cancer_genes()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_therapeutic_associated_genes" class="md-nav__link">
    get_therapeutic_associated_genes()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_genes_from_variant_types" class="md-nav__link">
    get_genes_from_variant_types()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_preferred_gene_name" class="md-nav__link">
    get_preferred_gene_name()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_cancer_predisposition_info" class="md-nav__link">
    get_cancer_predisposition_info()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_pharmacogenomic_info" class="md-nav__link">
    get_pharmacogenomic_info()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_gene_information" class="md-nav__link">
    get_gene_information()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="graphkbgenes">graphkb/genes</h1>
<p>Methods for retrieving gene annotation lists from GraphKB.</p>
<h2 id="get_oncokb_oncogenes">get_oncokb_oncogenes()</h2>
<p>Get the list of oncogenes stored in GraphKB derived from OncoKB.</p>
<pre><code class="language-python">def get_oncokb_oncogenes(conn: GraphKBConnection) -&gt; List[Ontology]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-ontology">Ontology</a>]: gene (Feature) records</li>
</ul>
<h2 id="get_oncokb_tumour_supressors">get_oncokb_tumour_supressors()</h2>
<p>Get the list of tumour supressor genes stored in GraphKB derived from OncoKB.</p>
<pre><code class="language-python">def get_oncokb_tumour_supressors(conn: GraphKBConnection) -&gt; List[Ontology]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-ontology">Ontology</a>]: gene (Feature) records</li>
</ul>
<h2 id="get_cancer_genes">get_cancer_genes()</h2>
<p>Get the list of cancer genes stored in GraphKB derived from OncoKB &amp; TSO500.</p>
<pre><code class="language-python">def get_cancer_genes(conn: GraphKBConnection) -&gt; List[Ontology]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-ontology">Ontology</a>]: gene (Feature) records</li>
</ul>
<h2 id="get_therapeutic_associated_genes">get_therapeutic_associated_genes()</h2>
<p>Genes related to a cancer-associated statement in Graphkb.</p>
<pre><code class="language-python">def get_therapeutic_associated_genes(graphkb_conn: GraphKBConnection) -&gt; List[Ontology]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>graphkb_conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-ontology">Ontology</a>]</li>
</ul>
<h2 id="get_genes_from_variant_types">get_genes_from_variant_types()</h2>
<p>Retrieve a list of Genes which are found in variants on the given types.</p>
<pre><code class="language-python">def get_genes_from_variant_types(
    conn: GraphKBConnection,
    types: List[str],
    source_record_ids: List[str] = [],
    ignore_cache: bool = False,
) -&gt; List[Ontology]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
<li>types (<code>List[str]</code>): list of names of variant types</li>
<li>source_record_ids (<code>List[str]</code>): list of sources ids to filter genes by</li>
<li>ignore_cache (<code>bool</code>)</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-ontology">Ontology</a>]: gene (Feature) records</li>
</ul>
<h2 id="get_preferred_gene_name">get_preferred_gene_name()</h2>
<p>Preferred gene symbol of a gene or transcript.</p>
<pre><code class="language-python">def get_preferred_gene_name(
    conn: GraphKBConnection, gene_name: str, source: str = PREFERRED_GENE_SOURCE
) -&gt; str:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
<li>gene_name (<code>str</code>): the gene name to search features by</li>
<li>source (<code>str</code>): id of the preferred gene symbol source</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>str</code>: preferred displayName symbol.</li>
</ul>
<p><strong>Examples</strong></p>
<pre><code class="language-python">return KRAS for get_preferred_gene_name(conn, 'NM_033360')
return KRAS for get_preferred_gene_name(conn, 'ENSG00000133703.11')
</code></pre>
<h2 id="get_cancer_predisposition_info">get_cancer_predisposition_info()</h2>
<p>Return two lists from GraphKB, one of cancer predisposition genes and one of associated variants.</p>
<p>GERO-272 - criteria for what counts as a "cancer predisposition" variant</p>
<p>In short:
* Statement 'source' is 'CGL'
* Statement 'relevance' is 'pathogenic'
* gene is gotten from any associated 'PositionalVariant' records</p>
<p>Example: https://graphkb.bcgsc.ca/view/Statement/155:11616</p>
<pre><code class="language-python">def get_cancer_predisposition_info(conn: GraphKBConnection) -&gt; Tuple[List[str], Dict[str, str]]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>Tuple[List[str], Dict[str, str]]</code>: list of cancer predisposition genes variants: dictionary mapping pharmacogenomic variant IDs to variant display names</li>
</ul>
<h2 id="get_pharmacogenomic_info">get_pharmacogenomic_info()</h2>
<p>Return two lists from GraphKB, one of pharmacogenomic genes and one of associated variants.</p>
<p>SDEV-2733 - criteria for what counts as a "pharmacogenomic" variant</p>
<p>In short:
* Statement 'source' is not 'CGI' or 'CIViC'
* Statement 'relevance' is 'increased toxicity' or 'decreased toxicity'
* gene is gotten from any associated 'PositionalVariant' records</p>
<p>Example: https://graphkb.bcgsc.ca/view/Statement/154:9574</p>
<pre><code class="language-python">def get_pharmacogenomic_info(conn: GraphKBConnection) -&gt; Tuple[List[str], Dict[str, str]]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>Tuple[List[str], Dict[str, str]]</code>: list of pharmacogenomic genes variants: dictionary mapping pharmacogenomic variant IDs to variant display names</li>
</ul>
<h2 id="get_gene_information">get_gene_information()</h2>
<p>Create a list of gene_info flag dicts for IPR report upload.</p>
<p>Function is originally from pori_ipr_python::annotate.py</p>
<p>Gene flags (categories) are: ['cancerRelated', 'knownFusionPartner', 'knownSmallMutation',
'oncogene', 'therapeuticAssociated', 'tumourSuppressor']</p>
<pre><code class="language-python">def get_gene_information(
    graphkb_conn: GraphKBConnection, gene_names: Sequence[str]
) -&gt; List[Dict[str, bool]]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>graphkb_conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): [description]</li>
<li>gene_names (<code>Sequence[str]</code>): [description]</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>List[Dict]</code>: <gene_str>, <flag>: True}] Keys of False values are simply omitted from ipr upload to reduce info transfer. eg. [{'cancerRelated': True, 'knownFusionPartner': True, 'knownSmallMutation': True, 'name': 'TERT', 'oncogene': True}]</li>
</ul>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
    
  </body>
</html>