
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.2, mkdocs-material-9.2.7">
    
    
      
        <title>graphkb/match - GraphKB Python</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.046329b4.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#graphkbmatch" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="GraphKB Python" class="md-header__button md-logo" aria-label="GraphKB Python" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GraphKB Python
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              graphkb/match
            
          </span>
        </div>
      </div>
    </div>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/bcgsc/pori_graphkb_python" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    github
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="GraphKB Python" class="md-nav__button md-logo" aria-label="GraphKB Python" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    GraphKB Python
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/bcgsc/pori_graphkb_python" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    github
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Notice
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#features_cache" class="md-nav__link">
    FEATURES_CACHE
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#get_equivalent_features" class="md-nav__link">
    get_equivalent_features()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cache_missing_features" class="md-nav__link">
    cache_missing_features()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#match_category_variant" class="md-nav__link">
    match_category_variant()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#match_copy_variant" class="md-nav__link">
    match_copy_variant()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#positions_overlap" class="md-nav__link">
    positions_overlap()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compare_positional_variants" class="md-nav__link">
    compare_positional_variants()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#type_screening" class="md-nav__link">
    type_screening()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#match_positional_variant" class="md-nav__link">
    match_positional_variant()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="graphkbmatch">graphkb/match</h1>
<p>Functions which return Variants from GraphKB which match some input variant definition</p>
<h2 id="features_cache">FEATURES_CACHE</h2>
<pre><code class="language-python">FEATURES_CACHE: Set[str] = set()
</code></pre>
<h2 id="get_equivalent_features">get_equivalent_features()</h2>
<p>Match an equivalent list of features given some input feature name (or ID).</p>
<pre><code class="language-python">def get_equivalent_features(
    conn: GraphKBConnection,
    gene_name: str,
    ignore_cache: bool = False,
    is_source_id: bool = False,
    source: str = &quot;&quot;,
    source_id_version: str = &quot;&quot;,
) -&gt; List[Ontology]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
<li>gene_name (<code>str</code>): the gene name to search features by</li>
<li>ignore_cache (<code>bool</code>): bypass the cache to always force a new request</li>
<li>is_source_id (<code>bool</code>): treat the gene_name as the gene ID from the source database (ex. ENSG001)</li>
<li>source (<code>str</code>): the name of the source database the gene definition is from (ex. ensembl)</li>
<li>source_id_version (<code>str</code>): the version of the source_id</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-ontology">Ontology</a>]: equivalent feature records</li>
</ul>
<p><strong>Examples</strong></p>
<pre><code class="language-python">get_equivalent_features(conn, 'KRAS')
</code></pre>
<pre><code class="language-python">get_equivalent_features(conn, 'ENSG001', source='ensembl', is_source_id=True)
</code></pre>
<pre><code class="language-python">get_equivalent_features(conn, 'ENSG001', source='ensembl', source_id_version='1')
</code></pre>
<pre><code class="language-python">get_equivalent_features(conn, '#3:44')
</code></pre>
<h2 id="cache_missing_features">cache_missing_features()</h2>
<p>Create a cache of features that exist to avoid repeatedly querying
for missing features</p>
<pre><code class="language-python">def cache_missing_features(conn: GraphKBConnection) -&gt; None:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
</ul>
<h2 id="match_category_variant">match_category_variant()</h2>
<p>Returns a list of variants matching the input variant</p>
<pre><code class="language-python">def match_category_variant(
    conn: GraphKBConnection,
    gene_name: str,
    category: str,
    root_exclude_term: str = &quot;&quot;,
    gene_source: str = &quot;&quot;,
    gene_is_source_id: bool = False,
    ignore_cache: bool = False,
) -&gt; List[Variant]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
<li>gene_name (<code>str</code>): the name of the gene the variant is in reference to</li>
<li>category (<code>str</code>): the variant category (ex. copy loss)</li>
<li>root_exclude_term (<code>str</code>)</li>
<li>gene_source (<code>str</code>): The source database the gene is defined by (ex. ensembl)</li>
<li>gene_is_source_id (<code>bool</code>): Indicates the gene name(s) input should be treated as sourceIds not names</li>
<li>ignore_cache (<code>bool</code>)</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-variant">Variant</a>]: List of variant records from GraphKB which match the input</li>
</ul>
<p><strong>Raises</strong></p>
<ul>
<li><a href="../util/#class-featurenotfounderror">FeatureNotFoundError</a>: The gene could not be found in GraphKB</li>
</ul>
<h2 id="match_copy_variant">match_copy_variant()</h2>
<p>Returns a list of variants matching the input variant</p>
<pre><code class="language-python">def match_copy_variant(
    conn: GraphKBConnection, gene_name: str, category: str, drop_homozygous: bool = False, **kwargs
) -&gt; List[Variant]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
<li>gene_name (<code>str</code>): the name of the gene the variant is in reference to</li>
<li>category (<code>str</code>): the variant category (ex. copy loss)</li>
<li>drop_homozygous (<code>bool</code>): Drop homozygous matches from the result when true</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-variant">Variant</a>]: List of variant records from GraphKB which match the input</li>
</ul>
<p><strong>Raises</strong></p>
<ul>
<li><code>ValueError</code>: The input copy category is not recognized</li>
</ul>
<h2 id="positions_overlap">positions_overlap()</h2>
<p>Check if 2 Position records from GraphKB indicate an overlap</p>
<pre><code class="language-python">def positions_overlap(
    pos_record: BasicPosition, range_start: BasicPosition, range_end: Optional[BasicPosition] = None
) -&gt; bool:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>pos_record (<a href="../types/#class-basicposition">BasicPosition</a>): the record to compare</li>
<li>range_start (<a href="../types/#class-basicposition">BasicPosition</a>): the position record indicating the start of an uncertainty range</li>
<li>range_end (Optional[<a href="../types/#class-basicposition">BasicPosition</a>]): the position record indicating the end of an uncertainty range</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>bool</code>: True if the positions overlap</li>
</ul>
<p><strong>Raises</strong></p>
<ul>
<li><code>NotImplementedError</code>: if a cytoband type position is given</li>
</ul>
<p>!!! note
    null values indicate not-specified or any</p>
<h2 id="compare_positional_variants">compare_positional_variants()</h2>
<p>Compare 2 variant records from GraphKB to determine if they are equivalent</p>
<pre><code class="language-python">def compare_positional_variants(
    variant: Union[PositionalVariant, ParsedVariant],
    reference_variant: Union[PositionalVariant, ParsedVariant],
    generic: bool = True,
) -&gt; bool:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>variant (Union[<a href="../types/#class-positionalvariant">PositionalVariant</a>, <a href="../types/#class-parsedvariant">ParsedVariant</a>]): the input variant</li>
<li>reference_variant (Union[<a href="../types/#class-positionalvariant">PositionalVariant</a>, <a href="../types/#class-parsedvariant">ParsedVariant</a>]): the reference (matched) variant record</li>
<li>generic (<code>bool</code>): also include the more generic variants</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>bool</code>: True if the records are equivalent</li>
</ul>
<h2 id="type_screening">type_screening()</h2>
<p>[KBDEV-1056]
Given a parsed variant notation, ensure that for some structural variant, type
(e.g. duplication, deletion, insertion, indel, copy number, inversion, etc.)
is only returned when the length of the variation meets a threshold,
otherwise 'mutation' is returned as default.</p>
<pre><code class="language-python">def type_screening(
    conn: GraphKBConnection,
    parsed: ParsedVariant,
    updateStructuralTypes=False,
) -&gt; str:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>): the graphkb connection object</li>
<li>parsed (<a href="../types/#class-parsedvariant">ParsedVariant</a>): the variant notation parsed as a dictionary by the API</li>
<li>updateStructuralTypes (<code>boolean</code>): if True the API is queried for an updated list</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li><code>str</code>: A string describing the variation type</li>
</ul>
<p><strong>Examples</strong></p>
<pre><code class="language-python"># structural variant type returned as 'mutation' IF length &lt; threshold (50)
type_screening(conn, {
'type': 'deletion',
'break1Start': {'pos': 1},
'break2Start': {'pos': 5},
}) -&gt; 'mutation'
</code></pre>
<pre><code class="language-python"># structural variant type returned as-is IF length &gt;= threshold (50)
type_screening(conn, {
'type': 'deletion',
'break1Start': {'pos': 1},
'break2Start': {'pos': 50},
}) -&gt; 'deletion'
</code></pre>
<pre><code class="language-python"># fusion &amp; translocation always returned as-is
type_screening(conn, {'type': 'fusion'}) -&gt; 'fusion'
</code></pre>
<pre><code class="language-python"># non structural always returned as-is
type_screening(conn, {'type': 'substitution'}) -&gt; 'substitution'
</code></pre>
<h2 id="match_positional_variant">match_positional_variant()</h2>
<p>Given the HGVS+ representation of some positional variant, parse it and match it to
annotations in GraphKB</p>
<pre><code class="language-python">def match_positional_variant(
    conn: GraphKBConnection,
    variant_string: str,
    reference1: Optional[str] = None,
    reference2: Optional[str] = None,
    gene_is_source_id: bool = False,
    gene_source: str = &quot;&quot;,
    ignore_cache: bool = False,
    updateStructuralTypes: bool = False,
) -&gt; List[Variant]:
</code></pre>
<p><strong>Args</strong></p>
<ul>
<li>conn (<a href="../util/#class-graphkbconnection">GraphKBConnection</a>)</li>
<li>variant_string (<code>str</code>): the HGVS+ annotation string</li>
<li>reference1 (<code>Optional[str]</code>): Explicitly specify the first reference link record (gene1)</li>
<li>reference2 (<code>Optional[str]</code>): Explicitly specify the second reference link record (gene2)</li>
<li>gene_is_source_id (<code>bool</code>): Indicates the gene name(s) input should be treated</li>
<li>gene_source (<code>str</code>): The source database the gene is defined by (ex. ensembl)</li>
<li>ignore_cache (<code>bool</code>)</li>
<li>updateStructuralTypes (<code>bool</code>): Whether or not updating the structural variant list</li>
</ul>
<p><strong>Returns</strong></p>
<ul>
<li>List[<a href="../types/#class-variant">Variant</a>]: A list of matched statement records</li>
</ul>
<p><strong>Raises</strong></p>
<ul>
<li><code>NotImplementedError</code>: thrown for uncertain position input (ranges)</li>
<li><a href="../util/#class-featurenotfounderror">FeatureNotFoundError</a>: One of the genes does not exist in GraphKB</li>
<li><code>ValueError</code>: the gene names were given both in the variant_string and explicitly</li>
</ul>
<p><strong>Examples</strong></p>
<pre><code class="language-python">match_positional_variant(conn, '(EWSR1,FLI1):fusion(e.1,e.2)')
</code></pre>
<pre><code class="language-python">match_positional_variant(conn, 'fusion(e.1,e.2)', 'EWSR1', 'FLI1')
</code></pre>
<pre><code class="language-python">match_positional_variant(conn, 'fusion(e.1,e.2)', '#3:4', '#4:5')
</code></pre>
<pre><code class="language-python">match_positional_variant(conn, 'fusion(e.1,e.2)', '123', '456', gene_is_source_id=True, gene_source='entrez gene')
</code></pre>
<pre><code class="language-python">match_positional_variant(conn, 'KRAS:p.G12D')
</code></pre>
<pre><code class="language-python">match_positional_variant(conn, 'p.G12D', 'KRAS')
</code></pre>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.dfff1995.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.dff1b7c8.min.js"></script>
      
    
  </body>
</html>